lib/modules.nix

```diff
@@ -430,10 +430,16 @@ rec {
 
       # an attrset 'name' => list of unmatched definitions for 'name'
       unmatchedDefnsByName =
-        # Propagate all unmatched definitions from nested option sets
-        mapAttrs (n: v: v.unmatchedDefns) resultsByName
-        # Plus the definitions for the current prefix that don't have a matching option
-        // removeAttrs defnsByName' (attrNames matchedOptions);
+        if configs == []
+        then
+          # When no config values exist, there can be no unmatched config, so
+          # we short circuit and avoid evaluating more _options_ than necessary.
+          {}
+        else
+          # Propagate all unmatched definitions from nested option sets
+          mapAttrs (n: v: v.unmatchedDefns) resultsByName
+          # Plus the definitions for the current prefix that don't have a matching option
+          // removeAttrs defnsByName' (attrNames matchedOptions);
```

Comment:
After taking a closer look, I think the main savings come from `defnsByName'` not being evaluated when `configs == []`, meaning that `config` sections aren't evaluated when not needed.

Since this does make module more lazy to some degree it would be awesome if we could have a test case that shows that ensures that extra lazy behavior so we can see the impact and prevent future regressions of it.

Reply:
I've tested this for options now, but not for config yet.